stdlib/public/core/StringGuts.swift

```diff
@@ -0,0 +1,1653 @@
+        internalDumpHex(start, newline: false)
+      } else {
+        print("<opaque>", terminator: "")
+      }
+      print(" count: ", terminator: "")
+      print(_nonTaggedCocoaCount, terminator: "")
+    } else if _isUnmanaged {
+      print("unmanaged ", terminator: "")
+      internalDumpHex(_unmanagedRawStart, newline: false)
+      print(" count: ", terminator: "")
+      print(_unmanagedCount, terminator: "")
+    } else if _isSmallCocoa {
+      print("small cocoa ", terminator: "")
+      internalDumpHex(_taggedCocoaObject, newline: false)
+      print(" count: ", terminator: "")
+      print(_taggedCocoaCount, terminator: "")
+    } else {
+      print("error", terminator: "")
+    }
+    if isASCII {
+      print(" <ascii>", terminator: "")
+    }
+    else {
+      print(" <utf16>", terminator: "")
+    }
+    print(")")
+  }
+}
+
+//
+// String API helpers
+//
+extension _StringGuts {
+  // Return a native storage object with the same contents as this string.
+  // Use the existing buffer if possible; otherwise copy the string into a
+  // new buffer.
+  @_versioned
+  internal
+  func _extractNativeStorage<CodeUnit>(
+    of codeUnit: CodeUnit.Type = CodeUnit.self
+  ) -> _SwiftStringStorage<CodeUnit>
+  where CodeUnit : FixedWidthInteger & UnsignedInteger {
+    if _fastPath(_isNative && CodeUnit.bitWidth == self.bitWidth) {
+      return _bridgeObject(toNative: _object) as! _SwiftStringStorage<CodeUnit>
+    }
+    let count = self.count
+    return _copyToNativeStorage(from: 0..<count, capacity: count)
+  }
+
+  @_specialize(where CodeUnit == UInt8)
+  @_specialize(where CodeUnit == UInt16)
+  @_specialize(where CodeUnit == UTF16.CodeUnit)
+  @_versioned
+  @_inlineable
+  internal
+  func _copyToNativeStorage<CodeUnit>(
+    of codeUnit: CodeUnit.Type = CodeUnit.self,
+    from range: Range<Int>,
+    capacity: Int = 0
+  ) -> _SwiftStringStorage<CodeUnit>
+  where CodeUnit : FixedWidthInteger & UnsignedInteger {
+    _sanityCheck(capacity >= 0)
+    let storage = _SwiftStringStorage<CodeUnit>.create(
+      capacity: Swift.max(range.count, capacity),
+      count: range.count)
+    self._copy(range: range, into: storage.usedBuffer)
+    return storage
+  }
+
+  @_inlineable
+  public // @testable
+  func _extractSlice(_ range: Range<Int>) -> _StringGuts {
+    if range.isEmpty { return _StringGuts() }
+    if range == 0..<count { return self }
+    switch (isASCII, _isUnmanaged) {
+    case (true, true):
+        return _StringGuts(_asUnmanaged(of: UInt8.self)[range])
+    case (true, false):
+      return _StringGuts(_copyToNativeStorage(of: UInt8.self, from: range))
+    case (false, true):
+      return _StringGuts(_asUnmanaged(of: UTF16.CodeUnit.self)[range])
+    case (false, false):
+      return _StringGuts(
+        _copyToNativeStorage(of: UTF16.CodeUnit.self, from: range))
+    }
+  }
+
+  @_inlineable
+  @inline(__always)
+  public // @testable
+  mutating func isUniqueNative() -> Bool {
```

Comment:
@lorentey. BTW, is this safe and sufficient in a multi-threaded environment? That is, do we instead need to ensure uniqueness for the entire execution of a method, or is that guaranteed / implied by the fact that self itself is local, and any copies would invariably bump the count? I don't know the rules here. CC @aschwaighofer @atrick @moiseev

Reply:
The standard library does not need to guard against language level races. A language level race is just that and needs to be addressed by the user of the standard library.

For example: 

‘’’
var a: String  // global

Thread 1:

a.mutate()

Thread 2:

a.mutate() // race on ‘a’
‘’’

Will race without any additional synchronization in the user program.